lib/accessly/policy/base.rb

```diff
@@ -0,0 +1,270 @@
+module Accessly
+  module Policy
+    class Base
+
+      attr_reader :actor
+
+      def initialize(actor)
+        @actor = actor
+      end
+
+      def self.actions(actions)
+        _actions.merge!(actions)
+        actions.each do |action, action_id|
+          _define_action_methods(action, action_id)
+        end
+      end
+
+      def self.actions_on_objects(actions_on_objects)
+        _actions_on_objects.merge!(actions_on_objects)
+        actions_on_objects.each do |action, action_id|
+          _define_action_methods(action, action_id)
+        end
+      end
+
+      def self.namespace
+        String(self)
+      end
+
+      def namespace
+        self.class.namespace
+      end
+
+      def self.model_scope
+        puts caller
+        raise ArgumentError.new("#model_scope is not defined on #{self.name}.")
+      end
+
+      def model_scope
+        self.class.model_scope
+      end
+
+      def is_admin?
+        false
+      end
+
+      def segment_id
+        nil
+      end
+
+      def grant(action, object = nil)
+        object_id = _get_object_id(object)
+
+        action_id = if object_id.nil?
+          _get_general_action_id!(action)
+        else
+          _get_action_on_object_id!(action)
+        end
+
+        grant_object.grant!(action_id, namespace, object_id)
+      end
+
+      def revoke(action, object = nil)
+        object_id = _get_object_id(object)
+
+        action_id = if object_id.nil?
+          _get_general_action_id!(action)
+        else
+          _get_action_on_object_id!(action)
+        end
+
+        revoke_object.revoke!(action_id, namespace, object_id)
+      end
+
+      def accessly_query
+        @_accessly_query ||= begin
+          query = Accessly::Query.new(actor)
```

Comment:
Is a policy strictly tied to an individual Actor? `Grant` and `Revoke` only accept an `actor` object, but `Query` can take the hash `{User => actor.id, Group => [1,3,5], Company => 9}` ... Do we want to provide a way in the policy to override `actor` for `Query` methods?  Do we expect that the developer will accomplish that in subclass policies?

Reply:
So ignore that ^ -- I see you added https://app.clubhouse.io/lessonly/story/9583/implement-groups-in-base-policy

Reply:
You got it!